### PR TITLE
Add HPO demo scripts

### PR DESCRIPTION
**Pull request by kusumachalasani**
Signed-off-by: kusumachalasani <kchalasa@redhat.com>

### CODE REVIEW

File: hpo_demo_setup.sh [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# Default docker image repos
+HPO_DOCKER_REPO="docker.io/kruize/hpo"
+HPO_VERSION="0.01"
```
Comment (dinogun) [inline hpo_demo_setup.sh:20]: We should not be hard coding versions here. Instead it should pick up the latest HPO version from some artifact in the HPO repo itself

File: hpo_demo_setup.sh [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+	echo "#######################################"
+	echo "Start a new experiment with search space json"
+	## Step 1 : Start a new experiment with provided search space.
+	exp_json=$( cat ${SEARCHSPACE_JSON} )
+	if [[ ${exp_json} != "" ]]; then
+		curl -Ss -H 'Content-Type: application/json' ${URL}/experiment_trials -d '{ "operation": "EXP_TRIAL_GENERATE_NEW",  "search_space": '"${exp_json}"'}'
+		check_err "Error: Creating the new experiment failed."
+	else
+		check_err "Error: Searchspace is empty"
+	fi
```
Comment (dinogun) [inline hpo_demo_setup.sh:151]: We can do away with the else here
```
if [ -z "${exp_json}" ]; then
    check_err "Error: Searchspace is empty"
fi
curl -Ss -H 'Content-Type: application/json' ${URL}/experiment_trials -d '{ "operation": "EXP_TRIAL_GENERATE_NEW",  "search_space": '"${exp_json}"'}'
check_err "Error: Creating the new experiment failed."
```

File: hpo_demo_setup.sh [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+		echo "#######################################"
+        	echo "Generate the config for trial ${i}"
+		sleep 2
+		HPO_CONFIG=$(curl -Ss -H 'Accept: application/json' "${URL}"'/experiment_trials?experiment_id='"${eid}"'&trial_number='"${i}")
+		#echo ${HPO_CONFIG}
+		if [[ ${HPO_CONFIG} != -1 ]]; then
+			echo "${HPO_CONFIG}" > hpo_config.json
+		else
+			check_err "Error: Issue generating the configuration from HPO."
+		fi
```
Comment (dinogun) [inline hpo_demo_setup.sh:166]: Same here, dont need the else
if [[ ${HPO_CONFIG} == -1 ]]; then
      check_err "Error: Issue generating the configuration from HPO."
fi
echo "${HPO_CONFIG}" > hpo_config.json

File: hpo_demo_setup.sh [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+
+		## Step 3: Run the benchmark with HPO config.
+		echo "#######################################"
+	        echo "Run the benchmark for trial ${i}"
+		BENCHMARK_OUTPUT=$(./hpo_helpers/runbenchmark.sh ${SEARCHSPACE_JSON})
+		#BENCHMARK_OUTPUT="Objfunc_result=0.007914818407446147 Benchmark_status=prune"
+		echo ${BENCHMARK_OUTPUT}
```
[the remark this answers is not in the document]
Comment (dinogun) [reply hpo_demo_setup.sh:173]: Also why is SEARCHSPACE_JSON being passed in here? Shouldn't it be hpo_config.json?

File: hpo_demo_setup.sh [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+		obj_result=$(echo ${BENCHMARK_OUTPUT} | cut -d "=" -f2 | cut -d " " -f1)
+		trial_state=$(echo ${BENCHMARK_OUTPUT} | cut -d "=" -f3 | cut -d " " -f1)
+		if [[ ${obj_result} == "" || ${trial_state} == "" ]]; then
+			obj_result=0
+			trial_state="prune"
+		fi
+		### Add the HPO config and output data from benchmark of all trials into single csv
+		${PY_CMD} -c "import hpo_helpers.utils; hpo_helpers.utils.hpoconfig2csv(\"hpo_config.json\",\"output.csv\",\"${exp_output}\",\"${i}\")"
```
Comment (dinogun) [inline hpo_demo_setup.sh:181]: This seems out of place here. Please create a function for this and move it there. This function can be called optionally to illustrate its usage.

**Comment by chandrams** [comment]
@kusumachalasani - I think we should install all the required modules through the script. 
```
./hpo_demo_setup.sh 

#######################################
#        HPO Demo Setup          #
#######################################


#######################################
 Cloning hpo git repos
done
#######################################


#######################################
 Start HPO Server

Starting hpo with  ./deploy_hpo.sh -c native

#######################################

Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/csubrama/github/kusuma/autotune-demo/hpo_helpers/utils.py", line 7, in <module>
    from datatest import validate
ModuleNotFoundError: No module named 'datatest'
#######################################
```

File: hpo_demo_setup.sh (hunk 1 above)
Comment (dinogun) [reply hpo_demo_setup.sh:21]: Also this should be HPO version `0.0.1` and not `0.01`

File: hpo_demo_setup.sh [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+
+# Default cluster
+CLUSTER_TYPE="native"
+PY_CMD="python3.6"
+LOGFILE="${PWD}/hpo.log"
+export N_TRIALS=3
+export N_JOBS=1
+
+function usage() {
+	echo "Usage: $0 [-s|-t] [-d] [-i hpo-image] [-r] [-c cluster-type]"
```
Comment (dinogun) [inline hpo_demo_setup.sh:31]: Please use `-o` for hpo as that is what we are using in all the repos now

File: hpo_demo_setup.sh [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+	pushd hpo >/dev/null
+		if [ -z "${HPO_DOCKER_IMAGE}" ]; then
+			HPO_DOCKER_IMAGE=${HPO_DOCKER_REPO}:${HPO_VERSION}
+		fi
+		if [[ ${CLUSTER_TYPE} == "native" ]]; then
+			echo
+			echo "Starting hpo with  ./deploy_hpo.sh -c ${CLUSTER_TYPE}"
+			echo
+			./deploy_hpo.sh -c ${CLUSTER_TYPE} >> ${LOGFILE}  &
+			#check_err "ERROR: HPO failed to start, exiting"
```
Comment (dinogun) [inline hpo_demo_setup.sh:118]: Why is this commented out

File: hpo_demo_setup.sh [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+		else
+			echo
+			echo "Starting hpo with  ./deploy_hpo.sh -c ${CLUSTER_TYPE} -h ${HPO_DOCKER_IMAGE}"
+			echo
+
+			./deploy_hpo.sh -c "${CLUSTER_TYPE}" -h "${HPO_DOCKER_IMAGE}"
+			#check_err "ERROR: HPO failed to start, exiting"
```
Comment (dinogun) [inline hpo_demo_setup.sh:125]: Why is this commented out?

File: hpo_demo_setup.sh [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+		fi
+	popd >/dev/null
+	echo "#######################################"
+	echo
+}
+
+###########################################
+#   Start HPO Experiments
+###########################################
```
Comment (dinogun) [inline hpo_demo_setup.sh:134]: Please add more comments here to indicate how this can be customized for other usecases

File: hpo_demo_setup.sh [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+		## Output format expected for BENCHMARK_OUTPUT is "Objfunc_result=0.007914818407446147 Benchmark_status=success"
+		echo "#######################################"
+		echo
+	        echo "Run the benchmark for trial ${i}"
+		echo
+		BENCHMARK_OUTPUT=$(./hpo_helpers/runbenchmark.sh "hpo_config.json" "${SEARCHSPACE_JSON}" "$i")
+		echo ${BENCHMARK_OUTPUT}
+		obj_result=$(echo ${BENCHMARK_OUTPUT} | cut -d "=" -f2 | cut -d " " -f1)
+		trial_state=$(echo ${BENCHMARK_OUTPUT} | cut -d "=" -f3 | cut -d " " -f1)
+		if [[ ${obj_result} == "" || ${trial_state} == "" ]]; then
```
Comment (dinogun) [inline hpo_demo_setup.sh:194]: Add a comment here to indicate what is considered as a bad result for pruning

File: hpo_demo_setup.sh [hunk 11]
```diff
@@ -0,0 +1,307 @@
...
+	end_time=$(get_date)
+	elapsed_time=$(time_diff "${start_time}" "${end_time}")
+	echo "Success! HPO demo cleanup took ${elapsed_time} seconds"
+	echo
+}
+
+# By default we start the demo and experiment
+hpo_restart=0
+start_demo=1
+HPO_DOCKER_IMAGE=""
```
Comment (dinogun) [inline hpo_demo_setup.sh:274]: We can set the default docker image for HPO by reading the version number in the file version.py in HPO repo

Comment (kusumachalasani) [reply hpo_demo_setup.sh:274]: I'm getting the version number in line 113 before starting the server if HPO_DOCKER_IMAGE is not set.

File: hpo_demo_setup.sh [hunk 12]
```diff
@@ -0,0 +1,310 @@
...
+LOGFILE="${PWD}/hpo.log"
+export N_TRIALS=3
+export N_JOBS=1
+
+function usage() {
+	echo "Usage: $0 [-s|-t] [-d] [-o hpo-image] [-r] [-c cluster-type]"
+	echo "s = start (default), t = terminate"
+	echo "r = restart hpo only"
+	echo "d = Don't start experiments"
+	echo "p = expose prometheus port"
```
Comment (dinogun) [inline hpo_demo_setup.sh:33]: 1. What are the valid cluster options ?
2. Is `d` option really used?
3. Why do we need prometheus port?

File: hpo_demo_setup.sh [hunk 13]
```diff
@@ -0,0 +1,310 @@
...
+		## Step 3: Run the benchmark with HPO config.
+		## Output of the benchmark should contain objective function result value and status of the benchmark.
+		## Status of the benchmark supported is success and prune
+		## Output format expected for BENCHMARK_OUTPUT is "Objfunc_result=0.007914818407446147 Benchmark_status=success"
+		## Status of benchmark trial is set to prune, if objective function result value is not a number.
+		echo "#######################################"
+		echo
+	        echo "Run the benchmark for trial ${i}"
+		echo
+		BENCHMARK_OUTPUT=$(./hpo_helpers/runbenchmark.sh "hpo_config.json" "${SEARCHSPACE_JSON}" "$i")
```
Comment (dinogun) [inline hpo_demo_setup.sh:188]: Looks like this benchmark requires openshift or a kubernetes cluster ?
If so do not start the experiments unless we are connected to a cluster

File: hpo_demo_setup.sh (hunk 12 above)
Comment (kusumachalasani) [reply hpo_demo_setup.sh]: Updated as required.

File: hpo_demo_setup.sh (hunk 13 above)
Comment (kusumachalasani) [reply hpo_demo_setup.sh:195]: The demo benchmark script uses only minikube to run. We are not giving options to user to give any parameters to run the benchmark.
I updated the usage to mention the cluster type in the demo script is only to start HPO service.

**Comment by dinogun** [comment]
Updated the PR with some pre-req checks. @kusumachalasani I suggest that you create a separate PR for the remaining pre-req checks. Will merge this for now

**Pull request merged**